File: JavaScript solutions/238. Product of Array Except Self.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var productExceptSelf = function(nums) {
    let pre = [];
    let suf = [];
    let ans = [];

    pre[0] = nums[0];
    for (let i = 1; i < nums.length; i++) {
        pre[i] = pre[i - 1] * nums[i];
    }

    suf[nums.length - 1] = nums[nums.length - 1];
    for (let i = nums.length - 2, j = 0; i >= 0; i--, j++) {
        suf[i] = suf[i + 1] * nums[i];
    }

    ans.push(suf[1]);
    for (let i = 1; i < nums.length - 1; i++) {
        ans.push(pre[i - 1] * suf[i + 1]);
    }
    ans.push(pre[nums.length - 2]);

    return ans;
};

File: C++ solutions/238. Product of Array Except Self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int>pre,suf,ans;
        pre.push_back(nums[0]);
        for(int i=1;i<nums.size();i++){
            pre.push_back(pre[i-1]*nums[i]);
        }

        suf.push_back(nums[nums.size()-1]);
        for(int i=nums.size()-2,j=0;i>=0;i--,j++){
            suf.push_back(suf[j]*nums[i]);
        }
        reverse(suf.begin(), suf.end());

        ans.push_back(suf[1]);
        for(int i=1;i<nums.size()-1;i++){
            ans.push_back(pre[i-1]*suf[i+1]);
        }
        ans.push_back(pre[pre.size()-2]);
        return ans;
    }
};